bindings/pyroot_experimental/PyROOT/src/DeclareCallablePyz.cxx

```diff
@@ -0,0 +1,663 @@
+      return NULL;
+   }
+
+   auto inputTypes = PyObject_GetAttrString(instance, "input_types");
+   auto returnType = PyObject_GetAttrString(instance, "return_type");
+
+   // Extract name of Python callable
+   PyObject* pyname;
+   if (PyObject_HasAttrString(instance, "name")) {
+      pyname = PyObject_GetAttrString(instance, "name");
+   } else {
+      if (!PyObject_HasAttrString(pyfunc, "__name__")) {
+         PyErr_SetString(PyExc_RuntimeError, "Failed to create C++ callable: Python callable does not have attribute __name__.");
+         return NULL;
+      }
+      pyname = PyObject_GetAttrString(pyfunc, "__name__");
+   }
+
+   auto name = CPyCppyy_PyUnicode_AsString(pyname);
+   Py_DECREF(pyname);
+
+   // Get C++ return type
+   if (!CPyCppyy_PyUnicode_Check(returnType)) {
+      PyErr_SetString(PyExc_RuntimeError, "Failed to create C++ callable: Return type argument cannot be interpreted as string.");
+      return NULL;
+   }
+   std::string returnTypeStr = CPyCppyy_PyUnicode_AsString(returnType);
+   Py_DECREF(returnType);
+   if (returnTypeStr.compare("") == 0) {
+      returnTypeStr = "void";
+   }
+
+   // Put function in GenericCallable namespace
+   std::stringstream code;
+   code << "namespace ROOT {\n";
+
+   // Set return type
+   code << returnTypeStr << " ";
+
+   // Set name of Python callable as function name
+   code << name;
+
+   // Build function signature, type string and list of variables
+   code << "(";
+
+   auto iter = PyObject_GetIter(inputTypes);
+   auto inputTypesSize = PyObject_Size(inputTypes);
+   Py_DECREF(inputTypes);
+   if (!iter) {
+      PyErr_SetString(PyExc_RuntimeError, "Failed to create C++ callable: Failed to iterate over input types.");
+      return NULL;
+   }
+
+   std::map<std::string, std::string> typemap = {
+       {"float", "f"},
+       {"double", "f"},
+       {"int", "i"},
+       {"unsigned int", "I"},
+       {"long", "l"},
+       {"unsigned long", "k"},
+   };
+
+   PyObject *item;
+   auto idx = 0u;
+   std::stringstream typestr;
+   std::vector<std::string> pytypes(inputTypesSize);
+   std::vector<std::string> inputTypesStr(inputTypesSize);
+   std::stringstream vars;
+   while ((item = PyIter_Next(iter))) {
+      // Convert argument to string
+      if (!CPyCppyy_PyUnicode_Check(item)) {
+         PyErr_SetString(PyExc_RuntimeError, "Failed to create C++ callable: Failed to interpret input type as string.");
+         return NULL;
+      }
+
+      inputTypesStr[idx] = CPyCppyy_PyUnicode_AsString(item);
+      Py_DECREF(item);
+
+      auto pytype = typemap.find(inputTypesStr[idx]);
+      if (pytype != typemap.end()) { // Types in typemap
+         pytypes[idx] = pytype->second;
+         vars << pytypes[idx] << "_" << idx;
+         code << inputTypesStr[idx] << " " << pytypes[idx] << "_" << idx;
+      } else if (inputTypesStr[idx].compare("") == 0) { // No input type
```

Comment:
Do we allow empty type names?

Reply:
If the signature takes no arguments, the list of types contains "empty" types.